### PR TITLE
feat(jmespath): add base types and errors

### PR DESCRIPTION
**Pull request by dreamorosi**
## Description of your changes

This PR adds the base components for the JMESPath utility. These includes constants, types, and error definitions.

The types include all the data structures used in the utility, most of which will be exposed to customers.

The error definitions include all the possible error types that can be thrown by the utility either during parsing, tokenization, or data extraction.

Finally, the constants includes the grammar for the JMESPath query language implemented [according to spec](https://jmespath.org/specification.html#grammar):
- Allowed characters/tokens/identifiers
- Supported digits and alphabet
- Binding power hash map - this hash map determines which tokens are processed together with others or as separate operations.

The PR also includes a stub for a module called `Functions` that will be added in a future PR.

### Related issues, RFCs

**Issue number:** #2193

## Checklist

- [x] [My changes meet the tenets criteria](https://docs.powertools.aws.dev/lambda/typescript/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [ ] I have made corresponding changes to the *documentation*
- [x] My changes generate *no new warnings*
- [ ] I have *added tests* that prove my change is effective and works
- [x] The PR title follows the [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

***Is it a breaking change?:*** NO

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by dreamorosi** [comment]
Hi @am29d, the docs workflow [is failing](https://github.com/aws-powertools/powertools-lambda-typescript/actions/runs/8230149315) because I made a mistake in the type doc config of this new utility.

Once this PR is merged the issue will go away.

**Review by am29d — APPROVED** [review]

**Comment by am29d** [comment]
Docs are still failing because of missing `README.md` file, which will be merged much later. To unblock the docs workflow, I'd remove the jmespath referecne in the API Docs for now

**Pull request merged**
